### PR TITLE
Added logging message for tasks that fail without error

### LINKED ISSUE #6633: Should still log a message even if `AllowFailureWithoutError` is set
Looking at this logic:

https://github.com/dotnet/msbuild/blob/e9593e841ab16e9792894267548e10b17c98c535/src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs#L937-L958

we should still log a message if `AllowFailureWithoutError` is set. If it's set, currently we log nothing, and that's makes it impossible to debug. We should log a message instead of logging nothing at all.

See here https://github.com/microsoft/vstest/issues/2953 for a real-world scenario where this would have helped a lot. VSTest sets AllowFailureWithoutError so we get ye olde

```
Build FAILED.
    0 Warning(s)
    0 Error(s)
```

### PR DESCRIPTION
**Pull request by edvilme**
Fixes #6633

### Changes Made
Added check for tasks that fail without error messages, so that a normal priority message is shown.

### Testing
Extended existing tests to validate new behavior. Also validated that in dotnet test scenarios, the new message is not visible.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=8125) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=8125) before we can accept your contribution.<br/><br/>:x: edvilme [sign now](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=8125)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/msbuild?pullRequest=8125) it.</sub>

**Comment by edvilme** [comment]
image
This is how the message gets logged in the binlog in a failing dotnet test.

File: global.json [hunk 1]
```diff
@@ -1,7 +1,8 @@
 {
-  "sdk": {
-    "allowPrerelease": true
-  },
+    "sdk": {
+        "allowPrerelease": true,
+        "version": "7.0.100-rc.2.22477.23"
+    },
```
Comment (rainersigwald) [inline global.json:5]: Sorry I didn't catch this! Can you please revert this change for this PR?

File: src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs [hunk 2]
```diff
@@ -961,11 +961,15 @@ private async Task<WorkUnitResult> ExecuteInstantiatedTask(ITaskExecutionHost ta
...
                     && (be is TaskHost th ? th.BuildRequestsSucceeded : false)
-                    && (be is IBuildEngine7 be7 ? !be7.AllowFailureWithoutError : true) // and it's not allowed to fail unless it logs an error
                     && !(_cancellationToken.CanBeCanceled && _cancellationToken.IsCancellationRequested)) // and it wasn't cancelled
                 {
+                    // If is allowed to fail without error
+                    if (be is IBuildEngine7 be7 && be7.AllowFailureWithoutError)
+                    {
+                        taskLoggingContext.LogComment(MessageImportance.Normal, "TaskReturnedFalseButDidNotLogError", _taskNode.Name);
+                    }
                     // Then decide how to log MSB4181
```
Comment (rainersigwald) [inline src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs:971]:
```suggestion
                    // Then decide how to log MSB4181
                    if (be is IBuildEngine7 be7 && be7.AllowFailureWithoutError)
                    {
                        // If it's allowed to fail without error, log as a message
                        taskLoggingContext.LogComment(MessageImportance.Normal, "TaskReturnedFalseButDidNotLogError", _taskNode.Name);
                    }

```

File: src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs [hunk 3]
```diff
@@ -961,11 +961,15 @@ private async Task<WorkUnitResult> ExecuteInstantiatedTask(ITaskExecutionHost ta
                     && !taskResult // and it returned false
                     && !taskLoggingContext.HasLoggedErrors // and it didn't log any errors
                     && (be is TaskHost th ? th.BuildRequestsSucceeded : false)
-                    && (be is IBuildEngine7 be7 ? !be7.AllowFailureWithoutError : true) // and it's not allowed to fail unless it logs an error
                     && !(_cancellationToken.CanBeCanceled && _cancellationToken.IsCancellationRequested)) // and it wasn't cancelled
                 {
+                    // If is allowed to fail without error
+                    if (be is IBuildEngine7 be7 && be7.AllowFailureWithoutError)
+                    {
+                        taskLoggingContext.LogComment(MessageImportance.Normal, "TaskReturnedFalseButDidNotLogError", _taskNode.Name);
```
Comment (rainersigwald) [inline src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs:969]: Fit-and-finish question (@baronfel, especially): This logs with the `MSB4181:` prefix. Generally, we try not to have "error codes" for message-level logging. But here, it's kind of a warning, if you squint really hard. Should we strip the prefix, or leave it as-is?

File: global.json (hunk 1 above)
Comment (edvilme) [reply global.json]: Sure! Removed now 👍🏻

File: src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs (hunk 3 above)
Comment (Forgind) [reply src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs:970]: The normal case is hopefully non-error (they rerouted the error(s) appropriately, so it's just an informational message they don't care about). Most of our other messages are similar to this in that they _might_ indicate an error but usually don't, so I'd vote for stripping the prefix, personally, but it's reasonable to leave it to baronfel.

**Review by Forgind — APPROVED** [review]

File: src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs (hunk 3 above)
Comment (rainersigwald) [reply src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs:970]: > Most of our other messages are similar to this in that they _might_ indicate an error but usually don't

I don't think I agree with this. We have a lot of messages that are fully neutral like "Started building target", and that feels like a very solid majority of logging in a normal build to me.

**Review by rainersigwald — APPROVED** [review]

File: src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs (hunk 3 above)
Comment (Forgind) [reply src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs:970]: I think of your example as "might indicate an error but usually don't"; if that target should not be building (or should not be building yet), it's indicating an error...but it's normally innocuous.
